### ISSUE TITLE
MapCache caching empty tiles when no nearest match is found

### ISSUE THREAD

**Issue by randomorder** [geosolutions-it/eumetsat-EOWS#484]
Matching JIRA issue [here](https://cgi-bds.atlassian.net/browse/AH-20)

After a rollout to OPE today we found that MapCache is caching empty tiles for WMS requests that are not matching nearest match value in GeoServer.

We've already faced the problem in the past and I think we had a patch in place for Eumetsat to overcome it. The trick was forcing GeoServer to return an exception to the client (MapCache in this case) in the event of no nearest match. I suspect the rollout of the improvement "[EUMETSAT - AH-6 - Nearest match error resolution](https://docs.google.com/document/d/1kj9GwHJvdmtvZKLGTxF-S_AFaAVVHq-4pwB2Lw7-41k/edit#)" / [#381](https://github.com/geosolutions-it/eumetsat-EOWS/issues/381) changed the behaviour and now blank tiles are being cached again.

As a temporary workaround I've asked EUM to prune empty tiles from MapCache. Can we harmonize the two and have AH-6 and still return an exception for no nearest match?

**Comment by aaime** [comment]
What seems we want here, is that when there is no nearest match:
- GS throws an exception if WMS request comes from MapCache
- GS does not throw an exception if the WMS request comes from the DownloadAnimation process

Ideally, it would be better if MapCache could avoid caching by recognizing the warning headers set by GeoServer. There is at least one warning about the lack of nearest match, not sure if WMS also adds "Cache-control: no cache" to the lot (if not, it would be easy to add).

If we have to go to the exceptions, we need to make this configurable. This fits into a larger context. When GeoServer is asked a time value that does not exist, it returns an empty image (has always done that) but the WMS specification says that an exception should be thrown instead. 
I would bet there are lots of clients out there that would break if we really did, but I guess we could make this behavior configurable. 
A request parameter could also be used to ask for exceptions in a more interactive way, in case they are disabled server side.

**Comment by aaime** [comment]
If we want a quick band-aid for their branch, I guess we can add a small bit of code in the part that adds the warnings in the HTTP headers, and throws an exception if the current request is a WMS one, rather than a WPS one. This should be doable in a matter of a few hours.

The larger plan described above requires a discussion on list, and some new settings in WMSInfo.

**Comment by simboss** [comment]
thx for the quick investigation @aaime.

Can you split this into two issues, one short term and one long term.
I would go ahead with both but we need the short term in the.... short term :)

**Comment by randomorder** [comment]
> I would bet there are lots of clients out there that would break if we really did, but I guess we could make this behavior configurable.
> A request parameter could also be used to ask for exceptions in a more interactive way, in case they are disabled server side.

Other projects would benefit too and we'd have the option to be compliant with the specification. 

For this specific project, injecting the param from MapCache is possible (we're already doing it to set TILED=false) but MapCache is only intercepting the `/wms` endpoint. We'd have an inconsistent behaviour for other clients using `/ows?SERVICE=WMS&...` and the other tile oriented services. I guess we could configure it globally to throw an exception instead to be consistent but what would happen to the WPS animation? 

For EUM we could configure it to throw an exception but what would happen to the WPS animations then? Should we ask for a small change to the client and pass a special parameter to avoid the exceptions?

Maybe I am overthinking it though @simboss

**Comment by aaime** [comment]
@randomorder I agree it's beneficial to have an option to be compliant, it just takes more time to develop and requires discussion. Hence the second story to work on it.

The behavior would not be inconsistent I believe, the code I intend to write would throw an exception no matter if the "/wms" or "/ows" are used, as long as an actual WMS request is made to the server, instead of an internal request made while running animations (where the current behavior is preserved).

That said, I'm not sure what MapCache does when it receives an exception? Does it throw it back?
We previously had an exception being thrown (that was breaking WPS animation), what was happening back then?

**Comment by randomorder** [comment]
> The behavior would not be inconsistent I believe, the code I intend to write would throw an exception no matter if the "/wms" or "/ows" are used, as long as an actual WMS request is made to the server, instead of an internal request made while running animations (where the current behavior is preserved).

I was referring to the use of extra param to get exceptions injected by MapCache. MapCache is only intercepting the `/wms` endpoint so for all requests using the `/wms` endpoint geoserver would receive the extra param and reply with an exception and for everything else we'd hit geoserver directly with no extra param and get empty tiles

> That said, I'm not sure what MapCache does when it receives an exception? Does it throw it back?
We previously had an exception being thrown (that was breaking WPS animation), what was happening back then?

If I remember correctly GS returned an exception to MC which in turn returned 502 to the clients. Ugly but it worked for EumetView

**Comment by randomorder** [comment]
short term fix deployed and tested
